### ISSUE TITLE
Gekko and Bollinger Band Strategies

### ISSUE THREAD

**Issue by raffus** [askmike/gekko#843]
Hi there,

I was wondering about how I could do a bollinger band strategy using Gekko.

Have anybody here ever thought about that? Is there a ready-to-go strategy using bollinger band?

BR,
R

**Comment by generalectric** [comment]
It's the same as combining any combination of Talib indicators in a strategy. Add the indicators then in check method just do if (candle.close > candle.bandup && candle.rsi > settings.rsi.up) for example

**Comment by raffus** [comment]
How can I enable talib?

When running gekko, I see this message

*TAlib is disabled*

2017-07-13 15:38 GMT-03:00 generalectric <notifications@github.com>:

**Comment by thegamecat** [comment]
Config, talib enable

Also npm install talib@1.0.2

**Comment by raffus** [comment]
Hey @thegamecat, thanks for the tip.
@generalectric, your help is much appreciated. Although I have no clue about how to implement this code. Is there a "bollinger" ready-to-go strategy?

I've installed talib but so far I can't tell if its working or not.

Thanks guys.

**Comment by raffus** [comment]
Any tip?
Thanks

**Comment by thegamecat** [comment]
in init
  this.addTalibIndicator('mybb', 'bbands', customBBSettings);

in check
  var result = this.talibIndicators.mybb.result;

  var lower = result['outRealLowerBand'];
  var medium = result['outRealMiddleBand'];
  var upper = result['outRealUpperBand'];

those are your bands.

**Comment by askmike** [comment]
Thanks @thegamecat, that is exactly how you can use bbands within Gekko.


@raffus if you need anything more than what @thegamecat posted feel free to keep asking. To answer your question:

> Is there a "bollinger" ready-to-go strategy?

The reason there is none is because of this:

> Bollinger Bands consist of three lines. The middle band is a simple moving average (generally 20 periods) of the typical price (TP). The upper and lower bands are F standard deviations (generally 2) above and below the middle band. The bands widen and narrow when the volatility of the price is higher or lower, respectively.
>
> **Bollinger Bands do not, in themselves, generate buy or sell signals; they are an indicator of overbought or oversold conditions.** When the price is near the upper or lower band it indicates that a reversal may be imminent. The middle band becomes a support or resistance level. The upper and lower bands can also be interpreted as price targets. When the price bounces off of the lower band and crosses the middle band, then the upper band becomes the price target.

TLDR: Bollinger Bands generate signals that can be seen as buy/sell singnals in a number of different ways.

**Comment by raffus** [comment]
Hey guys. Thanks a lot for the help.

After adding the code into my new strategy file, I'm getting this error

```
xxx POST /api/backtest 500 168ms -

  Error: non-error thrown: Child process has died.
      at Object.onerror (/Users/rafael/Documents/gekko/node_modules/koa/lib/context.js:105:40)
      at process._tickCallback (internal/process/next_tick.js:109:7)

```

This is my init function:

```
method.init = function() {
	// adding bollinger band
	this.addTalibIndicator('mybb', 'bbands', 'customBBSettings'); <<-- using '' between customBBSettings

	this.name = 'bollinger-test';
	this.historySize = config.tradingAdvisor.historySize;
}
```

And this is my check function:

```
method.check = function(candle) {
	var result = this.talibIndicators.mybb.result;
	var lower = result['outRealLowerBand'];
	var medium = result['outRealMiddleBand'];
	var upper = result['outRealUpperBand'];
}
```

If I put this into my init function:

```
method.init = function() {
	// adding bollinger band
	this.addTalibIndicator('mybb', 'bbands', customBBSettings); <<-- without ''

	this.name = 'bollinger-test';
	this.historySize = config.tradingAdvisor.historySize;
}
```
I get this error:

```
/Users/rafael/Documents/gekko/strategies/long.js:16
	this.addTalibIndicator('mybb', 'bbands', customBBSettings);
	                                         ^

ReferenceError: customBBSettings is not defined
    at Base.method.init (/Users/rafael/Documents/gekko/strategies/long.js:16:43)
    at Base.bound [as init] (/Users/rafael/Documents/gekko/node_modules/lodash/dist/lodash.js:729:21)
    at new Base (/Users/rafael/Documents/gekko/plugins/tradingAdvisor/baseTradingMethod.js:114:8)
    at Actor.setupTradingMethod (/Users/rafael/Documents/gekko/plugins/tradingAdvisor/tradingAdvisor.js:58:17)
    at Actor.bound [as setupTradingMethod] (/Users/rafael/Documents/gekko/node_modules/lodash/dist/lodash.js:729:21)
    at new Actor (/Users/rafael/Documents/gekko/plugins/tradingAdvisor/tradingAdvisor.js:23:8)
    at load (/Users/rafael/Documents/gekko/core/pluginUtil.js:95:22)
    at /Users/rafael/Documents/gekko/node_modules/async/dist/async.js:1156:9
    at replenish (/Users/rafael/Documents/gekko/node_modules/async/dist/async.js:1030:17)
    at iterateeCallback (/Users/rafael/Documents/gekko/node_modules/async/dist/async.js:1015:17)
  xxx POST /api/backtest 500 171ms -

  Error: non-error thrown: Child process has died.
      at Object.onerror (/Users/rafael/Documents/gekko/node_modules/koa/lib/context.js:105:40)
      at process._tickCallback (internal/process/next_tick.js:109:7)

```

Did I miss anything?

BR,
R

**Comment by askmike** [comment]
@raffus at this point I would recommend using the commandline to debug, if you use the UI you are wrapping a backtest in a lot of layers, one of which (the child process) eats the error output (stderr) so troubleshooting errors in your strategy become very hard.

Using the commandline is not that hard: you first copy the [sample-config.js](https://github.com/askmike/gekko/blob/stable/sample-config.js) to `config.js` and then:

- set [config.debug](https://github.com/askmike/gekko/blob/stable/sample-config.js#L10) to true
- set [tradingAdvisor.method](https://github.com/askmike/gekko/blob/stable/sample-config.js#L35) to the name of your strat (minus js).
- set the [candleSize and historySize](https://github.com/askmike/gekko/blob/stable/sample-config.js#L36-L37) as you would in the UI.
- add a config block for your strats parameters (if your strat requires it).

And finally run:

    node gekko --backtest --config config.js

Now you should see exactly what went wrong.

**Comment by raffus** [comment]
Hi @askmike. First I'd like to thank you for all your help.

After doing what you've told me, I still see the same error.

```
node gekko --backtest --config config.js

    ______   ________  __    __  __    __   ______
   /      \ /        |/  |  /  |/  |  /  | /      \
  /$$$$$$  |$$$$$$$$/ $$ | /$$/ $$ | /$$/ /$$$$$$  |
  $$ | _$$/ $$ |__    $$ |/$$/  $$ |/$$/  $$ |  $$ |
  $$ |/    |$$    |   $$  $$<   $$  $$<   $$ |  $$ |
  $$ |$$$$ |$$$$$/    $$$$$  \  $$$$$  \  $$ |  $$ |
  $$ \__$$ |$$ |_____ $$ |$$  \ $$ |$$  \ $$ \__$$ |
  $$    $$/ $$       |$$ | $$  |$$ | $$  |$$    $$/ 
   $$$$$$/  $$$$$$$$/ $$/   $$/ $$/   $$/  $$$$$$/

	Gekko v0.4.10
	I'm gonna make you rich, Bud Fox. 


2017-07-17 23:06:46 (INFO):	Setting up Gekko in backtest mode
2017-07-17 23:06:46 (INFO):	
2017-07-17 23:06:46 (INFO):	Setting up:
2017-07-17 23:06:46 (INFO):		 Trading Advisor
2017-07-17 23:06:46 (INFO):		 Calculate trading advice
2017-07-17 23:06:46 (INFO):		 Using the strategy: long
/Users/rafael/Documents/gekko/strategies/long.js:16
	this.addTalibIndicator('mybb', 'bbands', customBBSettings);
	                                         ^

ReferenceError: customBBSettings is not defined
    at Base.method.init (/Users/rafael/Documents/gekko/strategies/long.js:16:43)
    at Base.bound [as init] (/Users/rafael/Documents/gekko/node_modules/lodash/dist/lodash.js:729:21)
    at new Base (/Users/rafael/Documents/gekko/plugins/tradingAdvisor/baseTradingMethod.js:114:8)
    at Actor.setupTradingMethod (/Users/rafael/Documents/gekko/plugins/tradingAdvisor/tradingAdvisor.js:58:17)
    at Actor.bound [as setupTradingMethod] (/Users/rafael/Documents/gekko/node_modules/lodash/dist/lodash.js:729:21)
    at new Actor (/Users/rafael/Documents/gekko/plugins/tradingAdvisor/tradingAdvisor.js:23:8)
    at load (/Users/rafael/Documents/gekko/core/pluginUtil.js:95:22)
    at /Users/rafael/Documents/gekko/node_modules/async/dist/async.js:1156:9
    at replenish (/Users/rafael/Documents/gekko/node_modules/async/dist/async.js:1030:17)
    at iterateeCallback (/Users/rafael/Documents/gekko/node_modules/async/dist/async.js:1015:17)

```

Any clue?

Thanks a lot.

BR,
R

**Comment by raffus** [comment]
@askmike, I'm making some progress here.

I found out what the problem was:

We forgot to declare this variable:

```
var customBBSettings = {
		optInTimePeriod: 20,
		optInNbDevUp: 2,
		optInNbDevDn: 2,
		optInMAType: 2
	}
```
Which was called by the function:

`this.addTalibIndicator('mybb', 'bbands', customBBSettings);`

After adding this, the code ran without any error.

Now, I need help to figure out the correct parameters for Bollinger Band.

Would you know this information?

Thanks a lot!

**Comment by generalectric** [comment]
There are no correct parameters.  The parameters for desired results change with different candle sizes. But I find my bbands settings are always somewhere between 1 and 4. In log method add price and band debug outputs... Run a backtest and locate a point where it should be overbought or oversold then adjust to where price is higher or lower than the band. Bollinger bands alone are not the answer to perfect buy sell signals you will find you have to combine them with other indicators for optimal results. The good news is you added bbands to a strategy... Adding other indicators is a breeze... The rest is just if else statements meeting specific criteria.

**Comment by raffus** [comment]
Hi @generalectric. Thanks for this wonderful explanation. I've been learning a lot lately with Gekko. This seems to be a great "trading school".

As you said, bollinger bands itself are not perfect answer. Since I'm very new to this, would you like to recommend me another signal to work with it? What are your favorite strategies to combine with BBands?

After adding bollinger bands I'm still looking for better results. All the tests I've made are trending me to a complete failure, but I'm still optimist to overcome this challenge.

Just in time, I'm still having doubts about the parameters to use with bbands. I was unable to find any good documentation related to it. "optInMAType" is still a black-hole in my mind.

Thanks a lot.

**Comment by thegamecat** [comment]
The thing with indicators is you are really trying to recreate and optimise a trading strategy.

**Comment by endorianix** [comment]
any chance to get .js file for this bb strategy?

**Comment by xmusicxmakerx** [comment]
@raffus Combining B.B. with stochRSI is a really popular day trading strategy. Since you have B.B. added to your strategy, would you mind copy pasting the complete strategy so I can implement it as well. I have been looking for exactly what you're doing and have been unable to find anyone. I know approximately F-all about coding so gekko is really challenging for me. That being said I want to be able to use this tool to day trade instead of watching the market all the damn time lol

**Comment by deadheadtn** [comment]
@xmusicxmakerx  am almost done with BBStochRSI Strat.

**Comment by xmusicxmakerx** [comment]
Awesome! Thanks for the update. Excited about it.

**Comment by xmusicxmakerx** [comment]
Can I get an update on your B.B./stochRSI strat? Haha I’m excited to try and implement it.

**Comment by xmusicxmakerx** [comment]
@deadheadtn can I get an update on your B.B./stochRSI strat? Sorry forgot to tag you earlier. Ready to implement and backtest

**Comment by barrubba** [comment]
where i can find B.B./stochRSI strat ?
thqnk you

**Comment by MarianoL10** [comment]
Could I have please your B.B./stochRSI?
